app/src/main/java/com/example/translatorapp/screen/translate/TranslateFragment.kt

```diff
@@ -137,20 +140,51 @@ class TranslateFragment :
         }
 
         bind.translate.setOnClickListener {
-            translate(bind.input.text.toString().trim())
+            if (bind.input.text.toString().trim() != "") {
+                translateOnClick(bind.input.text.toString().trim())
+            }
+        }
+
+        bind.copyOutput.setOnClickListener {
+            (activity as MainActivity).copyToClipboard(bind.output.text.toString())
+        }
+
+        bind.copyInput.setOnClickListener {
+            (activity as MainActivity).copyToClipboard(bind.input.text.toString())
+        }
+
+        bind.inputSpeaker.setOnClickListener {
+            presenter.source?.code?.let {
+                (activity as MainActivity).speak(bind.input.text.toString(), Locale(it))
+            }
+        }
+
+        bind.outputSpeaker.setOnClickListener {
+            presenter.target?.code?.let {
+                val text = bind.output.text.toString().substringBefore("\n")
+                (activity as MainActivity).speak(text, Locale(it.substringBefore("-")))
+            }
         }
     }
 
-    private fun translate(text: String) {
+    private fun translateOnClick(text: String) {
         if (text.contains(" ")) {
             presenter.getTranslateSentence(text)
             presenter.breakSentence(text)
         } else {
             presenter.getTranslateWord(text)
         }
+        bind.copyInput.visibility = View.VISIBLE
+        bind.inputSpeaker.visibility = View.VISIBLE
+        bind.copyOutput.visibility = View.VISIBLE
+        bind.outputSpeaker.visibility = View.VISIBLE
     }
 
     private fun clearState() {
+        bind.copyInput.visibility = View.INVISIBLE
+        bind.inputSpeaker.visibility = View.INVISIBLE
+        bind.copyOutput.visibility = View.INVISIBLE
+        bind.outputSpeaker.visibility = View.INVISIBLE
```

Comment:
```suggestion
        bind.copyInput.visibility = View.VISIBLE
        bind.inputSpeaker.visibility = View.VISIBLE
        bind.copyOutput.visibility = View.VISIBLE
        bind.outputSpeaker.visibility = View.VISIBLE
    }

    private fun clearState() {
        bind.copyInput.visibility = View.INVISIBLE
        bind.inputSpeaker.visibility = View.INVISIBLE
        bind.copyOutput.visibility = View.INVISIBLE
        bind.outputSpeaker.visibility = View.INVISIBLE
```
chỗ này bị duplicate code, em tách ra thành 1 hàm truyền state vào và gọi nhé!